sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectExecutePlanHandler.scala

```diff
@@ -31,6 +32,18 @@ class SparkConnectExecutePlanHandler(responseObserver: StreamObserver[proto.Exec
     try {
       executeHolder.eventsManager.postStarted()
       executeHolder.start()
+    } catch {
+      // Errors raised before the execution holder has finished spawning a thread are considered
+      // plan execution failure, and the client should not try reattaching it afterwards.
+      case s: SparkThrowable =>
+        SparkConnectService.executionManager.removeExecuteHolder(executeHolder.key)
+        throw s
+      case t: Throwable =>
+        SparkConnectService.executionManager.removeExecuteHolder(executeHolder.key)
+        throw SparkException.internalError(t.getMessage(), t)
```

Comment:
Hmm, how would these exceptions arise from `executeHolder.start()` or `executeHolder.eventsManager.postStarted()`? These seem like simple operations that simply boot up other async process

Reply:
java.lang.OutOfMemoryError can be thrown *anywhere* heap memory is allocated (confirmed).

Ex:

```java.lang.OutOfMemoryError: null
	at java.lang.AbstractStringBuilder.hugeCapacity(AbstractStringBuilder.java:161) ~[?:1.8.0_392]
	at java.lang.AbstractStringBuilder.newCapacity(AbstractStringBuilder.java:155) ~[?:1.8.0_392]
	at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:125) ~[?:1.8.0_392]
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:472) ~[?:1.8.0_392]
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:486) ~[?:1.8.0_392]
	at java.lang.StringBuilder.append(StringBuilder.java:171) ~[?:1.8.0_392]
	at java.lang.StringBuilder.append(StringBuilder.java:81) ~[?:1.8.0_392]
	at grpc_shaded.com.google.protobuf.TextFormat$TextGenerator.print(TextFormat.java:856) ~[----ws_3_5--connector--connect--server--spark_connect_proto_lib_shaded-hive-2.3__hadoop-3.2_2.12--458243678--com.google.protobuf__protobuf-java__3.19.6.jar:?]
...
	at org.apache.spark.sql.connect.service.ExecuteEventsManager.postStarted(ExecuteEventsManager.scala:183) ~[----ws_3_5--connector--connect--server--spark_connect_proto_lib_shaded-hive-2.3__hadoop-3.2_2.12---1628521472--connect-prepare-hive-2.3__hadoop-3.2_2.12_deploy.jar:?]
	at org.apache.spark.sql.connect.service.SparkConnectExecutePlanHandler.handle(SparkConnectExecutePlanHandler.scala:32) ~[----ws_3_5--connector--connect--server--spark_connect_proto_lib_shaded-hive-2.3__hadoop-3.2_2.12---1628521472--connect-prepare-hive-2.3__hadoop-3.2_2.12_deploy.jar:?]
```

Reply:
You cannot wrap an OOM in a SparkException. Please use NonFatal here.

Reply:
Hm.. good point, yeah, OOM is definitely not a spark error.

On the other hand, OOM is definitely not NonFatal, so not sure if wrapping it in a NonFatal error type is a good idea. Unfortunately, not wrapping OOM is also problematic because (EDGE) the client will retry (OOM is translated into UNKNOWN) and will eventually get OPERATION_NOT_FOUND <- this isn't good either.

I'll firstly remove the second conversion, and address the client error issue (EDGE) separately.